main/persistent_settings.c

```diff
@@ -0,0 +1,200 @@
+  err = nvs_get_blob(nvs_handle, "config", out_config, &config_size);
+
+  if (err == ESP_ERR_NVS_NOT_FOUND) {
+    *out_config = persistent_settings_default;
+  } else {
+    ESP_RETURN_ON_ERROR(err, TAG, "Couldn't read value to NVS.");
+  }
+
+  nvs_close(nvs_handle);
+  return ESP_OK;
+}
+
+esp_err_t persistent_settings_get_timing_config(
+    enum can_bitrate_setting can_bitrate,
+    twai_timing_config_t *timing_config_out) {
+  switch (can_bitrate) {
+    case CAN_KBITS_25:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_25KBITS();
+      return ESP_OK;
+    case CAN_KBITS_50:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_50KBITS();
+      return ESP_OK;
+    case CAN_KBITS_100:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_100KBITS();
+      return ESP_OK;
+    case CAN_KBITS_125:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_125KBITS();
+      return ESP_OK;
+    case CAN_KBITS_250:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_250KBITS();
+      return ESP_OK;
+    case CAN_KBITS_500:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_500KBITS();
+      return ESP_OK;
+    case CAN_KBITS_800:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_800KBITS();
+      return ESP_OK;
+    case CAN_KBITS_1000:
+      *timing_config_out = (twai_timing_config_t)TWAI_TIMING_CONFIG_1MBITS();
+      return ESP_OK;
+    default:
+      return ESP_ERR_INVALID_ARG;
+  }
+}
+
+esp_err_t persistent_settings_setup_reset_button() {
+  button_config_t button_config = {
+      .type = BUTTON_TYPE_GPIO,
+      .long_press_time = 1000,
+      .gpio_button_config = {.gpio_num = 34, .active_level = 0}};
+  button_handle_t button_handle = iot_button_create(&button_config);
+  if (button_handle == NULL) {
+    ESP_LOGE(TAG, "Couldn't create button.");
+    return ESP_FAIL;
+  }
+
+  esp_err_t err = iot_button_register_cb(button_handle, BUTTON_LONG_PRESS_START,
+                                         button_handler, NULL);
+  ESP_RETURN_ON_ERROR(err, TAG, "Couldn't register button callback.");
+
+  return ESP_OK;
+}
+
+static void button_handler(void *button_handle, void *usr_data) {
+  ESP_LOGI(TAG, "Button 1 held. Resetting settings to default, and rebooting.");
+  ESP_ERROR_CHECK(persistent_settings_save(&persistent_settings_default));
+}
+
+esp_err_t persistent_settings_to_json(
+    const persistent_settings_t *current_config, char *buf_out, size_t buflen) {
+  int bytes_written = snprintf(
+      buf_out, buflen,
+      "{\n"
+      "\"eth_use_static\": "
+      "%s,\n"
+
+      "\"eth_ip\": "
+      "\"" IPSTR
+      "\",\n"
+
+      "\"eth_netmask\": "
+      "\"" IPSTR
+      "\",\n"
+
+      "\"eth_gw\": "
+      "\"" IPSTR
+      "\",\n"
+
+      "\"wifi_enabled\": "
+      "%s,\n"
+
+      "\"wifi_ssid\": "
+      "\"%s\",\n"
+
+      "\"wifi_pass\": "
+      "null,\n"
```

Comment:
I think it would be more obvious to return either "******" or something if there is a password and maybe "Not Set" or "" if not instead of always null.

Reply:
Fixed in 72219c3.